apps/aecore/lib/aecore/chain/block_validation.ex

```diff
@@ -3,115 +3,153 @@ defmodule Aecore.Chain.BlockValidation do
+           Chainstate.calculate_and_validate_chain_state(new_block, old_chain_state, height) do
       expected_root_hash = Chainstate.calculate_root_hash(new_chain_state)
 
-      expected_target =
-        Target.calculate_next_target(
-          time,
-          blocks_for_target_calculation
-        )
-
       cond do
         # do not check previous block height for genesis block, there is none
         !(is_genesis || check_correct_height?(new_block, previous_block)) ->
           {:error, "#{__MODULE__}: Incorrect height"}
 
+        !valid_prev_hash?(new_block, previous_block) ->
+          {:error, "#{__MODULE__}: Incorrect previous block hash"}
+
+        !valid_prev_key_hash?(new_block, previous_block) ->
+          {:error, "#{__MODULE__}: Incorrect previous key block hash"}
+
+        version != GovernanceConstants.protocol_version() ->
+          {:error, "#{__MODULE__}: Invalid protocol version"}
+
         !valid_header_time?(new_block) ->
           {:error, "#{__MODULE__}: Invalid header time"}
 
         root_hash != expected_root_hash ->
           {:error, "#{__MODULE__}: Root hash not matching"}
 
-        target != expected_target ->
-          {:error, "#{__MODULE__}: Invalid block target"}
-
         true ->
           {:ok, new_chain_state}
       end
     else
-      false ->
-        {:error, "#{__MODULE__}: One or more transactions don't meet their minimum fee"}
-
       {:error, _} = error ->
         error
     end
   end
 
-  @spec single_validate_block(Block.t()) :: :ok | {:error, String.t()}
-  def single_validate_block(
-        %Block{
-          header: %Header{txs_hash: txs_hash, version: version} = header,
-          txs: txs
-        } = block
-      ) do
-    block_txs_count = length(txs)
+  defp validate_key_block(
+         %KeyBlock{
+           header: %KeyHeader{target: target, time: time} = header
+         },
+         blocks_for_target_calculation
+       ) do
+    expected_target =
+      Target.calculate_next_target(
+        time,
+        blocks_for_target_calculation
+      )
 
     cond do
-      txs_hash != calculate_txs_hash(txs) ->
-        {:error, "#{__MODULE__}: Root hash of transactions does not match the one in header"}
+      target != expected_target ->
+        {:error, "#{__MODULE__}: Invalid block target"}
+
+      !is_solution_valid?(header) ->
+        {:error, "#{__MODULE__}: Invalid PoW solution"}
+
+      true ->
+        :ok
+    end
+  end
+
+  defp validate_micro_block(
+         %MicroBlock{
+           header:
+             %MicroHeader{time: new_time, signature: signature, txs_hash: txs_hash} = header,
+           txs: txs
+         },
+         %{
+           header: %{time: prev_block_time, prev_key_hash: prev_key_hash} = prev_header
+         } = prev_block
+       ) do
+    prev_key_block =
+      case prev_block do
+        %KeyBlock{} ->
+          prev_block
+
+        %MicroBlock{} ->
+          {:ok, key_block} = Chain.get_block(prev_key_hash)
```

Comment:
I think it would be better if you would pass `prev_key_block` as argument to this function and `calculate_and_validate_block`.
1. This is how we handle `prev_block`.
2. This way we don't do calls to Channel.Worker. (We remove circular dependency between models.)